darasimiolaifa/covid-19-estimator-darasimiolaifa
# feat(REST API): implement rest api for the estimator program

## What does this PR do?
This PR fulfills the solution to the challenge 3 of the BuildForSDG assessment test

## What are the major activities carried out?
- install express, typescript, xlm2js and their types
- create an express app with a route to calculate the estimate implemented in challenge 1-3
- create a route to allow user alternate between json and xml response format
- create route to allow request/response cycle duration to be logged
- write middleware to handle the writing of the logs

## How can this PR be manually tested?
- Clone the repo
- Install the dependencies
- Run `npm run watch`
- Launch Postman and send a `POST` request to `localhost:3000/api/v1/on-covid-19` with the following data:

```javascript

{
   region: {
     name: "Africa",
     avgAge: 19.7,
     avgDailyIncomeInUSD: 5,
     avgDailyIncomePopulation: 0.71
   },
   periodType: "days",
   timeToElapse: 58,
   reportedCases: 674,
   population: 66622705,
   totalHospitalBeds: 1380614
}
```
- You should get a response object that looks like this: (Notice the addition of the `impact` and `severeImpact` property objects)
```javascript

{
   data: {
      region: {
         name: "Africa",
         avgAge: 19.7,
         avgDailyIncomeInUSD: 5,
         avgDailyIncomePopulation: 0.71
       }
       periodType: "days",
       timeToElapse: 58,
       reportedCases: 674,
       population: 66622705,
       totalHospitalBeds: 1380614
   },
   impact: {
      currentlyInfected: 6740,
      infectionsByRequestedTime: 7067402240,
      severeCasesByRequestedTime: 1060110336,
      hospitalBedsByRequestedTime: -1059627121,
      casesForICUByRequestedTime: 176685056,
      casesForVentilatorByRequestedTime: 70674022,
      dollarsInFlight: 727589060607.9999
   },
   severeImpact: {
      currentlyInfected: 33700,
      infectionsByRequestedTime: 35337011200,
      severeCasesByRequestedTime: 5300551680,
      hospitalBedsByRequestedTime: -5300068465,
      casesForICUByRequestedTime: 883425280,
      casesForVentilatorByRequestedTime: 353370112,
      dollarsInFlight: 3637945303039.9995
   }
}
```
- Add `/json` to the url, and you should get the same response.
- Substitute the `/json` with `/xml`, and you should get this response:

```xml
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<root>
    <data>
       <region>
           <name>Africa</name>
           <avgAge>19.7</avgAge>
           <avgDailyIncomeInUSD>5</avgDailyIncomeInUSD>
           <avgDailyIncomePopulation>0.71</avgDailyIncomePopulation>
       </region>
       <periodType>days</periodType>
       <timeToElapse>58</timeToElapse>
       <reportedCases>674</reportedCases>
       <population>66622705</population>
       <totalHospitalBeds>1380614</totalHospitalBeds>
    </data>
    <impact>
        <currentlyInfected>6740</currentlyInfected>
        <infectionsByRequestedTime>3533701120</infectionsByRequestedTime>
        <severeCasesByRequestedTime>530055168</severeCasesByRequestedTime>
        <hospitalBedsByRequestedTime>-529571953</hospitalBedsByRequestedTime>
        <casesForICUByRequestedTime>176685056</casesForICUByRequestedTime>
        <casesForVentilatorsByRequestedTime>70674022</casesForVentilatorsByRequestedTime>
        <dollarsInFlight>216286878</dollarsInFlight>
    </impact>
    <severeImpact>
        <currentlyInfected>33700</currentlyInfected>
        <infectionsByRequestedTime>17668505600</infectionsByRequestedTime>
        <severeCasesByRequestedTime>2650275840</severeCasesByRequestedTime>
        <hospitalBedsByRequestedTime>-2649792625</hospitalBedsByRequestedTime>
        <casesForICUByRequestedTime>883425280</casesForICUByRequestedTime>
        <casesForVentilatorsByRequestedTime>353370112</casesForVentilatorsByRequestedTime>
        <dollarsInFlight>1081434394</dollarsInFlight>
    </severeImpact>
</root>
```
- Substitute the `/xml` with `/logs`, and send a `GET` request to the route, you should get a response similar to this:
```text
POST /api/v1/on-covid-19 200 1ms
POST /api/v1/on-covid-19/json 200 1ms
POST /api/v1/on-covid-19/xml 200 9ms 
```
